applications/vanilla/models/class.discussionmodel.php

```diff
@@ -1167,6 +1159,67 @@ public function calculate(&$discussion) {
         $this->fireEvent('SetCalculatedFields');
     }
 
+    /**
+     * Calculate if the user has read all the Comments in a discussion. The data in the Discussion table
+     * is more likely to be reliable because when Comments are deleted we update the Discussion table
+     * but not the UserDiscussion table.
+     *
+     * @param int $discussionCommentCount Number of Comments according to the Discussion table.
+     * @param mixed $discussionLastCommentDate Date of last Comment according to the Discussion table.
+     * @param int $userReadComments Number of Comments the user has read according to the UserDiscussion table.
+     * @param mixed $userLastReadDate Date of last Comment read according to the UserDiscussion table.
+     * @return array
+     */
+    public static function determineReadStatusAndUnreadCount(
+        int $discussionCommentCount,
+        $discussionLastCommentDate,
+        int $userReadComments,
+        $userLastReadDate
+    ) {
+        $isRead = true;
+        $unreadCommentCount = $discussionCommentCount - $userReadComments;
+        if ($discussionLastCommentDate > $userLastReadDate) {
+            $isRead = false;
+            if ($discussionCommentCount > 0 && $userReadComments >= $discussionCommentCount) {
+                $unreadCommentCount = 1;
+            }
+        }
+
+        if ($userLastReadDate > $discussionLastCommentDate && $unreadCommentCount > 0) {
+            $unreadCommentCount = 0;
+        }
+
+        if (!$userReadComments) {
+            $unreadCommentCount = $discussionCommentCount;
+        }
+
+        if ($unreadCommentCount < 0 && !$isRead) {
+            $unreadCommentCount = 1;
```

Comment:
You are missing coverage for this case.

I would also add comments to this and the next case. This case is a kludge and if you don't comment it another developer may remove it.

Reply:
It turns out that this case is actually already covered earlier in the method:
```
      if ($discussionLastCommentDate > $userLastReadDate) {
            $isRead = false;
            if ($discussionCommentCount > 0 && $userReadComments >= $discussionCommentCount) {
                $unreadCommentCount = 1;
            }
        }
```
However, we don't have a case for `if ($unreadCommentCount === 0 $$ !isRead)`

Is it ok to just return 0 for the unreadCount and false for $isRead?